onnxruntime/contrib_ops/cuda/bert/fast_gelu_impl.cu

```diff
@@ -0,0 +1,109 @@
+ https://github.com/NVIDIA/TensorRT/tree/release/5.1/demo/BERT/
+ 
+Copyright 2019 NVIDIA Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Modifications: Add (bias) before Gelu is merged into this op to get better performance.
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/shared_inc/cuda_call.h"
+#include <cuda_fp16.h>
+#include <cublas_v2.h>
+#include "fast_gelu_impl.h"
+
+using namespace onnxruntime::cuda;
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+// constants for approximating the normal cdf
+constexpr float A = 0.5;
+
+constexpr float B = 0.7978845608028654;  // sqrt(2.0/M_PI)
+
+constexpr float C = 0.035677408136300125;  // 0.044715 * sqrt(2.0/M_PI)
+
+__device__ inline float Tanh(const float& x) {
+  return tanhf(x);
+}
+
+__device__ inline half Tanh(const half& x) {
+  const float tmp = tanhf(__half2float(x));
+  return __float2half(tmp);
+}
+
+__device__ inline half2 Tanh(const half2& x) {
+  // at the moment, there is no half2 tanh builtin
+  float2 tmp = (__half22float2(x));
+  tmp.x = tanhf(tmp.x);
+  tmp.y = tanhf(tmp.y);
+  return __float22half2_rn(tmp);
+}
+
+template <typename T, unsigned TPB>
+__global__ void FastGeluKernel(const T a, const T b, const T c, int input_length, int bias_length, const T* input, const T* bias, T* output) {
+  const int idx = blockIdx.x * TPB + threadIdx.x;
+
+  if (idx < input_length) {
+    const T x = input[idx];
+    const T in = (bias == nullptr) ? x : (x + bias[idx % bias_length]);
+    const T cdf = a + a * Tanh(in * (c * in * in + b));
+    output[idx] = in * cdf;
+  }
+}
+
+template <>
+bool LaunchFastGeluKernel(cudaStream_t stream, int input_length, int bias_length, const float* input, const float* bias, float* output) {
+  constexpr int blockSize = 256;
+  const int gridSize = (input_length + blockSize - 1) / blockSize;
+  FastGeluKernel<float, blockSize><<<gridSize, blockSize, 0, stream>>>(A, B, C, input_length, bias_length, input, bias, output);
+
+  return CUDA_CALL(cudaPeekAtLastError());
+}
+
+template <>
+bool LaunchFastGeluKernel(cudaStream_t stream, int input_length, int bias_length, const half* input, const half* bias, half* output) {
+  constexpr int blockSize = 256;
+
+#if __CUDA_ARCH__ >= 530 || !defined(__CUDA_ARCH__)
```

Comment:
>__CUDA_ARCH__ [](start = 4, length = 13)

Does it work here? I think __CUDA_ARCH__ only works in device code.

Reply:
I think so. If it is removed, there is build error since some half2 functions not defined for old architecture.

Reply:
__CUDA_ARCH__ is undefined in host code
[https://stackoverflow.com/questions/16072740/the-behavior-of-cuda-arch-macro](https://stackoverflow.com/questions/16072740/the-behavior-of-cuda-arch-macro)

---
In reply to: [342718792](https://github.com/microsoft/onnxruntime/pull/2293#discussion_r342718792) [](ancestors = 342718792)

Reply:
Moved to device code.